modules/miner/update.go

```diff
@@ -1,11 +1,205 @@
+		m.setCounter++
+		m.fullSets[newSet.ID] = []int{m.setCounter}
+		var size uint64
+		var totalFees types.Currency
+		for i := range newSet.IDs {
+			size += newSet.Sizes[i]
+			for _, fee := range newSet.Transactions[i].MinerFees {
+				totalFees = totalFees.Add(fee)
+			}
+		}
+		// We will check to see if this splitSet belongs in the block.
+		s := &splitSet{
+			size:         size,
+			averageFee:   totalFees.Div64(size),
+			transactions: newSet.Transactions,
+		}
+
+		elem := &mapElement{
+			set:   s,
+			id:    splitSetID(m.setCounter),
+			index: 0,
+		}
+
+		m.splitSets[splitSetID(m.setCounter)] = s
+		newElements = append(newElements, elem)
+	}
+	return newElements
+}
+
+// addNewTxns adds new unconfirmed transactions to the miner's transaction
+// selection.
+func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	// Get new splitSets (in form of mapElement)
+	newElements := m.getNewSplitSets(diff)
+
+	// Place each elem in one of the MapHeaps
+	for _, elem := range newElements {
+		m.addMapElementTxns(elem)
+	}
+}
+
+// addMapElementTxns places the splitSet from a mapElement into the correct
+// mapHeap.
+func (m *Miner) addMapElementTxns(elem *mapElement) {
+	candidateSet := elem.set
+
+	// Check if heap for highest fee transactions has space.
+	if m.blockMapHeap.size+candidateSet.size < types.BlockSizeLimit-5e3 {
+		m.blockMapHeap.Push(elem)
+		return
+	}
+
+	// The block heap doesn't have enough space for this transaction.
+	// Check if removing sets from the blockMapHeap will be worth it.
+	// bottomSets will hold  the lowest fee sets from the blockMapHeap
+	bottomSets := make([]*mapElement, 0)
+	var sizeOfBottomSets uint64
+	var averageFeeOfBottomSets types.Currency
+
+	// While the heap cannot fit this set s, and while the (weighted)
+	// average fee for the lowest sets from the block is less than
+	// the fee for the set s, continue removing from the heap.
+	for {
+		_, exists := m.blockMapHeap.Peek()
+		// If the blockMapHeap is empty, push all elements removed from it
+		// back in, and place the candidate set into the overflow.
+		if !exists {
+			m.overflowMapHeap.Push(elem)
+			// Put back in transactions removed.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Go to the next candidate set.
+			break
+		}
+		nextSet := m.blockMapHeap.Pop()
+
+		bottomSets = append(bottomSets, nextSet)
+
+		// Calculating fees to compare total fee from those sets removed
+		// and the current set s.
+		totalFeeFromNextSet := nextSet.set.averageFee.Mul64(nextSet.set.size)
+		totalBottomFees := averageFeeOfBottomSets.Mul64(sizeOfBottomSets).Add(totalFeeFromNextSet)
+		sizeOfBottomSets += nextSet.set.size
+		averageFeeOfBottomSets := totalBottomFees.Div64(sizeOfBottomSets)
+
+		// If the average fee of the bottom sets from the block is higher
+		// than the fee from this candidate set, put the candidate into the
+		// overflow MapHeap.
+		if averageFeeOfBottomSets.Cmp(candidateSet.averageFee) == 1 {
```

Comment:
Things get convoluted I think if you focus primarily on the average fee of the bottom sets as a whole. Instead, you should allow it to be memoryless. Pop the lowest fees of a block until you either find a higher average fee for just the one set (at which point you give up), or until you have enough room to add the candidate set, and then you drop all the fees at once.

This way you can preserve the property that the lowest average fee in the block is higher than the highest average fee in the leftovers.